crates/broker/src/storage.rs

```diff
@@ -2,296 +2,311 @@
-                let data = tokio::fs::read(path)
-                    .await
-                    .map_err(|_| StorageErr::HttpStatusErr("File not found".to_string()))?;
-                Ok(data)
-            }
-            _ => Err(StorageErr::UnsupportedScheme(self.uri_scheme.clone())),
-        }
+impl Display for HttpHandler {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.url.fmt(f)
     }
+}
+
+#[async_trait]
+impl Handler for HttpHandler {
+    async fn fetch(&self) -> Result<Vec<u8>, StorageErr> {
+        let response = self
+            .client
+            .get(self.url.clone())
+            .send()
+            .await
+            .map_err(|err| StorageErr::Http(err.into()))?;
+        let response = response.error_for_status().map_err(|err| StorageErr::Http(err.into()))?;
+
+        // If a maximum size is set and the content_length exceeds it, return early.
+        let capacity_hint = response.content_length().unwrap_or_default() as usize;
+        if capacity_hint > self.max_size {
+            return Err(StorageErr::SizeLimitExceeded(capacity_hint));
+        }
+
+        let mut buffer = Vec::with_capacity(capacity_hint);
+        let mut stream = response.bytes_stream();
 
-    pub fn id(&self) -> Result<String, StorageErr> {
-        match self.uri_scheme.as_ref() {
-            "bonsai" => Ok(self.uri.path()[1..].to_string()),
-            _ => Err(StorageErr::UnsupportedScheme(self.uri_scheme.clone())),
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk.map_err(|err| StorageErr::Http(err.into()))?;
+            buffer.extend_from_slice(chunk.chunk());
+            if buffer.len() > self.max_size {
+                return Err(StorageErr::SizeLimitExceeded(buffer.len()));
+            }
         }
+
+        Ok(buffer)
     }
 }
 
-#[derive(Default)]
-pub struct UriHandlerBuilder {
-    uri_str: String,
-    max_size: Option<usize>,
-    retries: Option<u8>,
-    cache_dir: Option<PathBuf>,
+pub struct S3Handler {
+    bucket: String,
+    key: String,
+    client: S3Client,
+    max_size: usize,
 }
 
-impl UriHandlerBuilder {
-    pub fn new(uri_str: &str) -> Self {
-        Self { uri_str: uri_str.into(), max_size: None, retries: None, cache_dir: None }
-    }
+impl S3Handler {
+    async fn new(
+        url: url::Url,
+        max_size: usize,
+        max_retries: Option<u8>,
+        endpoint_url: Option<String>,
+        aws_iam_role_arn: Option<String>,
+    ) -> Result<Self, StorageErr> {
+        let retry_config = if let Some(max_retries) = max_retries {
+            RetryConfig::standard().with_max_attempts(max_retries as u32 + 1)
+        } else {
+            RetryConfig::disabled()
+        };
 
-    pub fn set_max_size(mut self, max_size: usize) -> Self {
-        self.max_size = Some(max_size);
-        self
-    }
+        let mut s3_conf = aws_config::from_env().retry_config(retry_config).load().await;
```

Comment:
This loads the AWS credentials from a file where we could document this somehow.

Reply:
I don't follow, it loads it from env variables doesnt it?

Reply:
Oops, sorry, obviously ment loaded from env in the comment.

Reply:
Agreed we do need to document this somewhere though. Perhaps @sashaaldrick has a suggestion

Reply:
I have an issue for this here: https://linear.app/risczero/issue/BM-756/create-docs-for-kms-encrypted-partner-inputs